### PR TITLE
Tow 13 14 15 16 user auth

### PR DESCRIPTION
**Pull request by JordanBandur**
## JIRA Ticket Link
TOW-13
https://atriadev.atlassian.net/browse/TOW-13?atlOrigin=eyJpIjoiYTgyMTczYTRkN2Q0NDZhZTkxMjYyZjAzOGVlYWFmOGYiLCJwIjoiaiJ9
TOW-14
https://atriadev.atlassian.net/browse/TOW-14?atlOrigin=eyJpIjoiMzE0NTliZjE0MDUyNGM5Nzg0NmRlODhhNWRhNmYzNTAiLCJwIjoiaiJ9
TOW-15
https://atriadev.atlassian.net/browse/TOW-15?atlOrigin=eyJpIjoiODQxYmMxNWEyZjMyNDRhM2IxZmE0YWNhZWI4OTQxNTUiLCJwIjoiaiJ9
TOW-16
https://atriadev.atlassian.net/browse/TOW-16?atlOrigin=eyJpIjoiMWMyYjdjNjBkNTAyNDA3NmEyN2E4YWViYzU3NDMwM2UiLCJwIjoiaiJ9

## Migrations Required?
YES :white_check_mark:

## Summary (a few sentences describing what this PR is doing) 
Adds several new service methods in the VolunteerServices class to enhance volunteer authentication and password management functionality. The key updates include:

1. authenticate_volunteer Method: Implements volunteer authentication using Django's built-in authentication system, including error handling, logging, and rate limiting to prevent brute force attacks.

2. validate_username_and_password Method: Validates the format and security of a volunteer's username and password before authentication, ensuring they meet the required criteria.

3. encrypt_password Method: Securely hashes a volunteer's password before saving it to the database, using Django's password hashing functions.

4. change_password Method: Facilitates secure password changes for volunteers by validating the old password, enforcing password policies, and sending notifications after a successful change.

## Checks
-  [ :white_check_mark:] I have ran all the tests locally, and they all pass

## Impacted Areas in Application
List the areas in the app this PR is going to affect 
* models.py - Volunteer
* dao.py - Volunteer
* services.py - New service methods for Volunteer
* test_volunteer.py

## Learnings
User authentication with python/django

### CODE REVIEW

File: townhall/myapi/services.py [hunk 1]
```diff
@@ -37,6 +50,135 @@ def update_volunteer(update_volunteer_data: UpdateVolunteerData) -> None:
...
+        :param username: The volunteer's username
+        :param password: The volunteer's password
+        :return: Volunteer object if authentication is successful, None otherwise
+        """
+        # Rate limiting
+        rate_limit_key = f'login_attempts_{username}'
+        attempts = cache.get(rate_limit_key, 0)
+        if attempts >= 5:  # Assuming a limit of 5 attempts
+            logger.warning(f'Rate limit exceeded for username: {username}')
+            raise ValidationError(_('Too many login attempts. Please try again later.'))
```
Comment (ansel-rangers11) [inline townhall/myapi/services.py:67]: This is pretty cool

File: townhall/myapi/services.py [hunk 2]
```diff
@@ -37,6 +50,135 @@ def update_volunteer(update_volunteer_data: UpdateVolunteerData) -> None:
...
+            if not volunteer.is_active:
+                logger.warning(f'Inactive account: {username}')
+                raise ValidationError(_('Account is inactive.'))
+            
+            logger.info(f'Successful login for username: {username}')
+            cache.delete(rate_limit_key)  # Reset on successful login
+            return volunteer
+        else:
+            logger.warning(f'Failed login attempt for username: {username}')
+            cache.set(rate_limit_key, attempts + 1, timeout=300)  # 5-minute timeout
```
Comment (ansel-rangers11) [inline townhall/myapi/services.py:81]: Cool timeout

File: townhall/myapi/services.py [hunk 3]
```diff
@@ -37,6 +50,135 @@ def update_volunteer(update_volunteer_data: UpdateVolunteerData) -> None:
...
+
+        :param volunteer_id: The volunteer's ID
+        :param old_password: The volunteer's current password
+        :param new_password: The volunteer's new password
+        :raises AuthenticationError: If the old password is incorrect
+        :raises ValidationError: If the new password fails validation or is a recent password
+        """
+        volunteer = volunteer_dao.get_volunteer(id=volunteer_id)
+        if not volunteer:
+            logger.error(f'Volunteer not found with ID: {volunteer_id}')
```
Comment (ansel-rangers11) [inline townhall/myapi/services.py:146]: Good use of logs

File: townhall/myapi/services.py [hunk 4]
```diff
@@ -37,6 +50,135 @@ def update_volunteer(update_volunteer_data: UpdateVolunteerData) -> None:
...
+        encrypted_password = VolunteerServices.encrypt_password(new_password)
+        volunteer.password = encrypted_password
+        volunteer.save()
+
+        # Update recent passwords cache
+        recent_passwords = (recent_passwords[-4:] if len(recent_passwords) >= 5 else recent_passwords) + [new_password]
+        cache.set(f'recent_passwords_{volunteer_id}', recent_passwords, timeout=None)
+
+        # Send notification email
+        send_mail(
```
Comment (ansel-rangers11) [inline townhall/myapi/services.py:174]: Good use of Django modules

File: townhall/tests/test_volunteer.py [hunk 5]
```diff
@@ -135,56 +131,147 @@ def test_get_all_volunteers_empty(self, MockVolunteerServices):
...
+        MockVolunteerServices.change_password(volunteer_id=1, old_password="OldPassword123!", new_password="NewPassword456!")
+
+        # Assert: Ensure the change password method was called with correct arguments
+        MockVolunteerServices.change_password.assert_called_once_with(volunteer_id=1, old_password="OldPassword123!", new_password="NewPassword456!")
+
+    def test_change_password_failure_due_to_old_password(self, MockVolunteerServices):
+        # Arrange: Mock failure due to incorrect old password
+        MockVolunteerServices.change_password.side_effect = ValidationError("Old password is incorrect")
+
+        # Act & Assert: Attempt to change password and expect a ValidationError
```
Comment (ansel-rangers11) [inline townhall/tests/test_volunteer.py:263]: Great tests in AAA

**Review by ansel-rangers11 — APPROVED** [review]
I learned a lot from reviewing this PR and did not realize Django has these powerful authenthication methods at our disposal, just a heads up to regenerate the migration if other PRs have updated the migration. I think it at least needs to be 0009 and 0010 (but could be 0010 and 0011 if the other PR is merged)

**Review by GagenvirGill — APPROVED** [review]

**Pull request merged**